src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,714 @@
+ * connection object.
+ */
+type NodeConnectionMap = Map<string, ConnectionAndLock>;
+
+/**
+ * This interface is provided to each `NodeConnection` in the connection map
+ * to provide necessary features. The main one being `killSelf` to allow the
+ * `NodeConnection` to tell the `NodeConnectionManager` to destroy the
+ * connection thus removing it from  the connection map.
+ */
+export interface NodeConnectionManagerInterface {
+  withConnF: NodeConnectionManager['withConnF'];
+  getSeedNodes: NodeConnectionManager['getSeedNodes'];
+  sendHolePunchMessage: NodeConnectionManager['sendHolePunchMessage'];
+  destroyCallback(): Promise<void>;
+}
+
+interface NodeConnectionManager extends StartStop {}
+@StartStop()
+class NodeConnectionManager {
+  /**
+   * Time used to estalish `NodeConnection`
+   */
+  public readonly connConnectTime: number;
+
+  /**
+   * Time to live for `NodeConnection`
+   */
+  public readonly connTimeoutTime: number;
+  /**
+   * Alpha constant for kademlia
+   * The number of closest nodes to contact initially
+   */
+  public readonly initialClosestNodes: number;
+
+  protected logger: Logger;
+  protected nodeGraph: NodeGraph;
+  protected keyManager: KeyManager;
+  protected fwdProxy: ForwardProxy;
+  protected revProxy: ReverseProxy;
+  protected seedNodes: SeedNodes;
+  protected connections: NodeConnectionMap = new Map();
+
+  public constructor({
+    keyManager,
+    nodeGraph,
+    fwdProxy,
+    revProxy,
+    seedNodes = {},
+    initialClosestNodes = 3,
+    connConnectTime = 20000,
+    connTimeoutTime = 60000,
+    logger,
+  }: {
+    nodeGraph: NodeGraph;
+    keyManager: KeyManager;
+    fwdProxy: ForwardProxy;
+    revProxy: ReverseProxy;
+    seedNodes?: SeedNodes;
+    initialClosestNodes?: number;
+    connConnectTime?: number;
+    connTimeoutTime?: number;
+    logger?: Logger;
+  }) {
+    this.logger = logger ?? new Logger(NodeConnectionManager.name);
+    this.keyManager = keyManager;
+    this.nodeGraph = nodeGraph;
+    this.fwdProxy = fwdProxy;
+    this.revProxy = revProxy;
+    this.seedNodes = seedNodes;
+    this.initialClosestNodes = initialClosestNodes;
+    this.connConnectTime = connConnectTime;
+    this.connTimeoutTime = connTimeoutTime;
+  }
+
+  public async start() {
+    this.logger.info(`Starting ${this.constructor.name}`);
+    for (const nodeIdEncoded in this.seedNodes) {
+      const nodeId = nodesUtils.decodeNodeId(nodeIdEncoded)!;
+      await this.nodeGraph.setNode(nodeId, this.seedNodes[nodeIdEncoded]);
+    }
+    this.logger.info(`Started ${this.constructor.name}`);
+  }
+
+  public async stop() {
+    this.logger.info(`Stopping ${this.constructor.name}`);
+    for (const [nodeId, connAndLock] of this.connections) {
+      if (connAndLock == null) continue;
+      if (connAndLock.connection == null) continue;
+      // It exists so we want to destroy it
+      await this.destroyConnection(IdInternal.fromString<NodeId>(nodeId));
+    }
```

Comment:
Like here when it iterates over the `this.connections`. You can just do `await this.destoryConnection(nodeId);` instead.

Reply:
Ok once this is changed to `NodeIdString`, you will need to cast the string form. But probably not during `IdInternal.fromString`, only during `toString`. Like `nodeId.toString() as NodeIdString`.

Reply:
Not applicable since we're using NodeIdString for the key now.